azurerm/internal/services/springcloud/spring_cloud_custom_domain_resource.go

```diff
@@ -0,0 +1,151 @@
+package springcloud
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/appplatform/mgmt/2020-07-01/appplatform"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/springcloud/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/springcloud/validate"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceSpringCloudCustomDomain() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSpringCloudCustomDomainCreateUpdate,
+		Read:   resourceSpringCloudCustomDomainRead,
+		Update: resourceSpringCloudCustomDomainCreateUpdate,
+		Delete: resourceSpringCloudCustomDomainDelete,
+
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.SpringCloudCustomDomainID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.SpringCloudCustomDomainName,
+			},
+
+			"spring_cloud_app_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.SpringCloudAppID,
+			},
+
+			"certificate_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"thumbprint": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
```

Comment:
I could changed it to `forcenew`, but I could not understand the reason. I have tried to modify the name of `azurerm_spring_cloud_certificate` to let it recreate, it actually failed and reoprts: `The cert xxx is used by domain xx.xx.xx, can not be deleted`. This behaviour is controlled by the backend service, I think change it to `forcenew` does not help.

what's more, if this field is `forcenew`, I think maybe field `certificate_name` should also be `forcenew` ?

another thought, since `certificate_name` and `thumbprint` both comes from `azurerm_spring_cloud_certificate`, maybe I could directly reference the resource id

Reply:
> The cert xxx is used by domain xx.xx.xx, can not be deleted
Correct, this is exactly the behaviour I was suggesting we avoid. If a user updates / replaces the referenced certificate, Terraform will be unable to delete that referenced certificate and will fail apply, since it will be in use.  By having `ForceNew` on this field (which will have a different value for a replace certificate, where name won't necessarily), we recreate the Custom Domain to release the association and allow the underlying certificate to be replaced. Does that make sense?

> maybe I could directly reference the resource id
I think this would be too restrictive?  The name may not be the only one on the certificate and another name on the same certificate (and subsequently same thumbprint) could be changed without needing to recreate this virtual resource? If it's a difference certificate, then the thumbprint will change and trigger the `ForceNew` in any case.

Reply:
thanks you for your explanation. I have made it `forcenew`